src/libraries/Common/src/System/Reflection/Metadata/TypeNameParserHelpers.cs

```diff
@@ -0,0 +1,747 @@
+            {
+                int offset = startIndex;
+                for (; offset >= 0; offset--)
+                {
+                    if (fullName[offset] is '.' or '+')
+                    {
+                        if (offset == 0 || fullName[offset - 1] != EscapeCharacter)
+                        {
+                            break;
+                        }
+                        offset--; // skip the escaping character
+                    }
+                }
+                return offset;
+            }
+        }
+
+        internal static string GetRankOrModifierStringRepresentation(int rankOrModifier)
+        {
+            return rankOrModifier switch
+            {
+                ByRef => "&",
+                Pointer => "*",
+                SZArray => "[]",
+                1 => "[*]",
+                _ => ArrayRankToString(rankOrModifier)
+            };
+
+            static string ArrayRankToString(int arrayRank)
+            {
+                Debug.Assert(arrayRank >= 2 && arrayRank <= 32);
+
+#if NET8_0_OR_GREATER
+                return string.Create(2 + arrayRank - 1, arrayRank, (buffer, rank) =>
+                {
+                    buffer[0] = '[';
+                    for (int i = 1; i < rank; i++)
+                        buffer[i] = ',';
+                    buffer[^1] = ']';
+                });
+#else
+                ValueStringBuilder sb = new(stackalloc char[16]);
+                sb.Append('[');
+                for (int i = 1; i < arrayRank; i++)
+                    sb.Append(',');
+                sb.Append(']');
+                return sb.ToString();
+#endif
+            }
+        }
+
+        /// <summary>
+        /// Are there any captured generic args? We'll look for "[[" and "[" that is not followed by "]", "*" and ",".
+        /// </summary>
+        internal static bool IsBeginningOfGenericAgs(ref ReadOnlySpan<char> span, out bool doubleBrackets)
+        {
+            doubleBrackets = false;
+
+            if (!span.IsEmpty && span[0] == '[')
+            {
+                // There are no spaces allowed before the first '[', but spaces are allowed after that.
+                ReadOnlySpan<char> trimmed = TrimStart(span.Slice(1));
+                if (!trimmed.IsEmpty)
+                {
+                    if (trimmed[0] == '[')
+                    {
+                        doubleBrackets = true;
+                        span = TrimStart(trimmed.Slice(1));
+                        return true;
+                    }
+                    if (!(trimmed[0] is ',' or '*' or ']')) // [] or [*] or [,] or [,,,, ...]
+                    {
+                        span = trimmed;
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        internal static ReadOnlySpan<char> TrimStart(ReadOnlySpan<char> input) => input.TrimStart();
```

Comment:
Why is this needed?

Reply:
TLDR: It was a leftover, I am going to remove it.

The initial safe parser prototype from was trimming only spaces, the CLR parser was trimming all whitespaces. I've removed the latter and tests were passing, I've found a test gap and sent https://github.com/dotnet/runtime/pull/98562 with all kinds of whitespaces. Then I've changed it from `TrimStart(' ')` to `TrimStart`, but have not inlined the helper method.